<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <meta http-equiv="X-UA-Compatible" content="ie=edge" />

    <title>ATOMDANCE-molecular-dynamics</title>
<!--
DREAM PULSE
https://templatemo.com/tm-536-dream-pulse
-->
    <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Open+Sans:400,600" />
    <link rel="stylesheet" href="css/all.min.css" />
    <link rel="stylesheet" href="css/bootstrap.min.css" />
    <link rel="stylesheet" href="slick/slick.css"/>    
    <link rel="stylesheet" href="slick/slick-theme.css"/>    
    <link rel="stylesheet" href="css/magnific-popup.css">
    <link rel="stylesheet" href="css/templatemo-dream-pulse.css" />
  </head>
  <body>
    <main class="container-fluid">
      <div class="row">        
          <nav id="tmSidebar" class="tm-bg-black-transparent tm-sidebar">
            <button class="navbar-toggler" type="button" aria-label="Toggle navigation">
              <i class="fas fa-bars"></i>
            </button>
            <div class="tm-sidebar-sticky">
              <div class="tm-brand-box">
                <div class="tm-double-border-1">
                  <div class="tm-double-border-2">
                    <h1 class="tm-brand text-uppercase">ATOM DANCE</h1>
                  </div>
                </div>
              </div>

              <ul id="tmMainNav" class="nav flex-column text-uppercase text-right tm-main-nav">
                <li class="nav-item">
                  <a href="#intro" class="nav-link active">
                    <span class="d-inline-block mr-3">Intro</span>
                    <span class="d-inline-block tm-white-rect"></span>
                  </a>
                </li>
                <li class="nav-item">
                  <a href="#about" class="nav-link">
                    <span class="d-inline-block mr-3">About</span>
                    <span class="d-inline-block tm-white-rect"></span>
                  </a>
                </li>
                <li class="nav-item">
                  <a href="#work" class="nav-link">
                    <span class="d-inline-block mr-3">Examples</span>
                    <span class="d-inline-block tm-white-rect"></span>
                  </a>
                </li>
                <li class="nav-item">
                  <a href="#clients" class="nav-link">
                    <span class="d-inline-block mr-3">Download</span>
                    <span class="d-inline-block tm-white-rect"></span>
                  </a>
                </li>
                <li class="nav-item">
                  <a href="#talk" class="nav-link">
                    <span class="d-inline-block mr-3">Documentation</span>
                    <span class="d-inline-block tm-white-rect"></span>
                  </a>
                </li>
              </ul>
              <ul class="nav flex-row tm-social-links">
                <li class="nav-item">
                  <a href="https://facebook.com" class="nav-link tm-social-link">
                    <i class="fab fa-facebook-f"></i>
                  </a>
                </li>
                <li class="nav-item">
                  <a href="https://twitter.com" class="nav-link tm-social-link">
                    <i class="fab fa-twitter"></i>
                  </a>
                </li>
                <li class="nav-item">
                  <a href="https://dribbble.com" class="nav-link tm-social-link">
                    <i class="fab fa-dribbble"></i>
                  </a>
                </li>
                <li class="nav-item">
                  <a href="https://linkedin.com" class="nav-link tm-social-link">
                    <i class="fab fa-linkedin-in"></i>
                  </a>
                </li>
              </ul>
              <footer class="text-center text-white small">
                <p class="mb--0 mb-2">Copyright 2020 Dream Pulse</p>
                <p class="mb-0">Design:
                  <a rel="nofollow" href="https://templatemo.com" class="tm-footer-link">Template Mo</a>
                </p>
              </footer>
            </div>
          </nav>
          
          <main role="main" class="ml-sm-auto col-12">
            <div
              class="parallax-window"
              data-parallax="scroll"
              data-image-src="img/header.jpg">
              <div class="tm-section-wrap">
                <section id="intro" class="tm-section">
                    <div class="tm-bg-white-transparent tm-intro">
                        <h2 class="tm-section-title mb-5 text-uppercase tm-color-gray">Introducing ATOMDANCE v1</h2>
                        <p class="tm-color-gray">
                          a python-based pipeline/plugin for statistical machine learning assisted comparison of molecular dynamic trajectories of proteins in two functional states (e.g. mutant vs. wild-type or bound vs unbound).
                        </p>
                        <p class="mb-0 tm-color-gray">
                          ATOMDANCE is developed by Dr Gregory A. Babbitt (Bioinformatics) and Dr. Ernest P Fokoue (mathematics) and graduate and undergraduate students at the Rochester Institute of Technology in Rochester, NY USA.
                        </p>
                    </div>              
                </section>
            </div>            
          </div>

          <div class="tm-section-wrap bg-white">
            <section id="about" class="row tm-section">
              
              <div class="tm-w-80">
                          <h2 class="tm-color-primary tm-section-title mb-4">ATOMDANCE v 1.3 - comparative molecular dynamics</h2>
                          <p class="mb-5">ATOMDANCE software containing DROIDS 5.0/maxDemon 4.0/Choreograph 2.0 is a python-based suite of machine learning assisted statistical methods for comparing molecular dynamic trajectories of proteins in two functional states (e.g. unbound vs. bound to something or wildtype vs mutated or hot vs. cold). It was developed on a python 3 science stack and only additionally requires the cpptraj library software and UCSF Chimerax molecular visualization software to be installed. The methods and software is offered freely (without guarantee) under GPL 3.0 and was developed by Dr. Gragory A. Babbitt, Dr. Ernest P. Fokoue and bioinformatics students at the Rochester Institute of Technology in 2017-2023.</p>
                          <h2 class="tm-color-primary tm-section-title mb-4">ATOMDANCE combines 4 main programs</h2>
                          <p>DROIDS 5.0 - (Detecting Relative Outlier Impacts in Dynamic Simulation) providing stie-wise comparisons (i.e. divergence metrics) and hypothesis testing for amino acid fluctuations</p>
                          <p>maxDemon 4.0 - (kernel-based machine learning for comparative protein dynamics) providing (A) site-wise comparisons of atom fluctuations and atom correlations utilizing max mean discrepancy (MMD) on learned features and (B) site-wise identification of non-neutral evolutionary changes in molecular dynamics (also via MMD).
                          <p>Choreograph 2.0 - multi-model ANOVA designed for pair-wise identification of coordinated site dynamics (i.e. resonance) as well as pair-wise atom fluctuation similarities between adjacent and non-adjacent sites (i.e. potential contact sites across chains)</p>
              </div>
              <div>
                <p>We also offer an easy user interface for running the machine learning post processor</p>
              </div>
              <img src="img/atomdance_gui.png" alt="Client Image" class="img-fluid tm-client-img" />
              <div>
                <p>We also offer a user interface for running MD simulations</p>
              </div>
              <img src="img/MDgui.png" alt="Client Image" class="img-fluid tm-client-img" />
              
              <div class="tm-w-80">
                          <div><i class="fas fa-5x fa-university mb-5 tm-section-icon"></i></div>
                          <h2 class="tm-color-primary tm-section-title mb-4">citations</h2>
                          <p class="mb-5">
                      <p class="tm-color-primary">ATOMDANCE PREPRINT</p>
				                <p class="mb-0">  Babbitt G.A. et al. 2023. ATOMDANCE: machine learning denoising and resonance analysis for functional and evolutionary comparisons of protein dynamics bioRxiv 2023.04.20.537698; doi: https://doi.org/10.1101/2023.04.20.537698
				                <a rel="nofollow" href="pub/preprint.pdf" style="background-color: green; color: white;">.pdf_</a></p><p></p>
                        <p></p>
                      <p class="tm-color-primary">PUBLISHED SOFTWARE NOTES AND THEORY </p>
				                <p class="mb-0"> Babbitt G.A. Coppola E.E. Mortensen J.S. Adams L.E. Liao J. K. 2018. DROIDS 1.2 â€“ a GUI-based pipeline for GPU-accelerated comparative protein dynamics. BIOPHYSICAL JOURNAL 114: 1009-1017. CELL Press. 
				                <a rel="nofollow" href="pub/droids1.pdf" style="background-color: green; color: white;">.pdf_</a></p><p></p>
                        <p class="mb-0"> Babbitt G.A. Fokoue E. Evans J.R. Diller K.I. Adams L.E. 2020. DROIDS 3.0 - Detection of genetic and drug class variant impact on conserved protein binding dynamics. BIOPHYSICAL JOURNAL 118: 541-551 CELL Press.
				                <a rel="nofollow" href="pub/droids2.pdf" style="background-color: green; color: white;">.pdf_</a></p><p></p>
				                <p class="mb-0"> Babbitt G.A. Fokoue E.P. Srivastava H.R. Callahan B. Rajendran M. 2022. Statistical machine learning for comparative protein dynamics with the DROIDS/maxDemon software pipeline. STAR PROTOCOLS 3(1):101194
				                <a rel="nofollow" href="pub/droids3.pdf" style="background-color: green; color: white;">.pdf_</a></p><p></p>
				              <p></p>
				              <p class="tm-color-primary">PUBLISHED APPLICATIONS IN CANCER AND INFECTIOUS DISEASE </p>
				                <p class="mb-0"> Rajendran M. Ferran M.C. Mouli L. Babbitt G.A. Lynch M.L. 2023. Evolution of drug resistance drives destabilization of flap region dynamics in HIV-1 protease. BIOPHYSICAL REPORTS 3(3):100121. 
				                <a rel="nofollow" href="pub/app6.pdf" style="background-color: green; color: white;">.pdf_</a></p><p></p>
				                <p class="mb-0"> Rajendran M.  Babbitt G.A. 2022. Persistent cross-species SARS-CoV-2 variant infectivity predicted via comparative molecular dynamics simulation. ROYAL SOCIETY OPEN SCIENCE 9(11):220600. 
				                <a rel="nofollow" href="pub/app5.pdf" style="background-color: green; color: white;">.pdf_</a></p><p></p>
				                <p class="mb-0"> Rajendran M. Ferran M.C. Babbitt G.A. 2022. Identifying vaccine escape sites via statistical comparisons of short-term molecular dynamics.  BIOPHYSICAL REPORTS  2(2):100056 
				                <a rel="nofollow" href="pub/app4.pdf" style="background-color: green; color: white;">.pdf_</a></p><p></p>
				                <p class="mb-0"> Rynkiewicz P. Babbitt G.A. Cui F. Hudson A.O. Lynch M.L. 2021. Functional binding dynamics relevant to the evolution of zoonotic spillovers in endemic and emergent Betacoronavirus strains. JOURNAL OF BIOMOLECULAR STRUCTURE AND DYNAMICS. DOI: 10.1080/07391102.2021.1953604 
				                <a rel="nofollow" href="pub/app3.pdf" style="background-color: green; color: white;">.pdf_</a></p><p></p>
				                <p class="mb-0"> Babbitt G.A. Lynch M. McCoy. M. Fokoue E.P. Hudson A.O. 2020. Function and evolution of B-Raf loop dynamics relevant to cancer recurrence under drug inhibition. JOURNAL OF BIOMOLECULAR STRUCTURE AND DYNAMICS doi: 10.1080/07391102.2020.1815578 
				                <a rel="nofollow" href="pub/app2.pdf" style="background-color: green; color: white;">.pdf_</a></p><p></p>
				  
                    </p>
              </div>  
                
                
              
            </section>
          </div>
            
          <div class="tm-section-wrap bg-white">
            <section id="work" class="row tm-section">
              <div class="col-12">
                <div class="w-100 tm-double-border-1 tm-border-gray">
                    <div class="tm-double-border-2 tm-border-gray tm-box-pad">
                      <div class="tm-gallery-wrap">
                          <h2 class="tm-color-primary tm-section-title mb-4 ml-2">Some Illustrative Example Analyses</h2>
                          <div class="tm-gallery">
                            <div class="tm-gallery-item">                        
                              <figure class="effect-bubba">
                                <img src="img/gallery/01.jpg" alt="Our Work Image" class="img-fluid">
                                <figcaption>
                                  <h2> <span>DNA-protein interaction</span></h2>
                                  <p>binding site ID via MMD on atom fluctuation</p>
                                  <a href="img/gallery/large/01.jpg">View more</a>
                                </figcaption>
                              </figure>
                            </div>
                            <div class="tm-gallery-item">
                              <figure class="effect-bubba">
                                <img src="img/gallery/02.jpg" alt="Our Work Image" class="img-fluid">
                                <figcaption>
                                  <h2><span>protein-drug interaction</span></h2>
                                  <p>BRAF activation loop ID via MMD on atom fluctuation</p>
                                  <a href="img/gallery/large/02.jpg">View more</a>
                                </figcaption>
                              </figure>
                            </div>
                            <div class="tm-gallery-item">
                              <figure class="effect-bubba">
                                <img src="img/gallery/03.jpg" alt="Our Work Image" class="img-fluid">
                                <figcaption>
                                  <h2><span>virus-target interaction</span></h2>
                                  <p>ID sites via MMD on atom fluctuation</p>
                                  <a href="img/gallery/large/03.jpg">View more</a>
                                </figcaption>
                              </figure>
                            </div>
                            <div class="tm-gallery-item">
                              <figure class="effect-bubba">
                                <img src="img/gallery/04.jpg" alt="Our Work Image" class="img-fluid">
                                <figcaption>
                                  <h2><span>human-bacterial ortholog</span></h2>
                                  <p>evolutionary impacts on dynamics</p>
                                  <a href="img/gallery/large/04.jpg">View more</a>
                                </figcaption>
                              </figure>
                            </div>
                            <div class="tm-gallery-item">
                              <figure class="effect-bubba">
                                <img src="img/gallery/05.jpg" alt="Our Work Image" class="img-fluid">
                                <figcaption>
                                  <h2><span>resonance analysis</span></h2>
                                  <p>for ID of coordinated site dynamics</p>
                                  <a href="img/gallery/large/05.jpg">View more</a>
                                </figcaption>
                              </figure>
                            </div>
                            <div class="tm-gallery-item">
                              <figure class="effect-bubba">
                                <img src="img/gallery/06.jpg" alt="Our Work Image" class="img-fluid">
                                <figcaption>
                                  <h2><span>alternative plots</span></h2>
                                  <p>for KL divergence metrics</p>
                                  <a href="img/gallery/large/06.jpg">View more</a>
                                </figcaption>
                              </figure>
                            </div>
                            <div class="tm-gallery-item">
                              <figure class="effect-bubba">
                                <img src="img/gallery/07.jpg" alt="Our Work Image" class="img-fluid">
                                <figcaption>
                                  <h2><span>alternative plots</span></h2>
                                  <p>for amino acid ID, bootstrap p-value, RMSD over time</p>
                                  <a href="img/gallery/large/07.jpg">View more</a>
                                </figcaption>
                              </figure>
                            </div>
                          </div>
                      </div>                        
                    </div>                  
                  </div>     
              </div>         
            </section>
          </div>
            
          <div class="tm-section-wrap bg-white">
            <section id="clients" class="row tm-section">
              <div class="col-12 tm-section-pad">
                  <div class="tm-flex-item-left">
                      <button onclick="window.location.href='https://github.com/gbabbitt/ATOMDANCE-comparative-protein-dynamics';">GitHub repository</button>
                      <div><p></p></div>
                      <div class="tm-w-80">
                          <h2 class="tm-color-primary tm-section-title mb-4">Download and License</h2>
                          <p class="mb-5">
                            ATOMDANCE is offered freely without warranty or guarantee under a GPL 3.0 license and can be found at the GitHub repo linked above.  If you publish using our work, please cite us using one or more of the papers listed above
                          </p>
                      <button onclick="window.location.href='pub/mandoc.pdf';">ATOMDANCE user guide</button>
                      </div>
                      <div><p></p></div>
                      <div class="tm-w-80">
                          <h2 class="tm-color-primary tm-section-title mb-4">Documentation</h2>
                          <p class="mb-5">
                            We offer an extensive user guide .pdf in the link above.  This guide covers installation, explanation of methods, user instructions, and example output.
                          </p>
                      
                      </div>
                      
                  </div> 
              </div>                             
            </section>
          </div>
            
          
        </main>        
      </div>
    </div>
    <script src="js/jquery.min.js"></script>
    <script src="js/jquery.singlePageNav.min.js"></script>
    <script src="js/parallax.min.js"></script>
    <script src="slick/slick.min.js"></script>
    <script src="js/jquery.magnific-popup.min.js"></script>
    <script src="js/templatemo-scripts.js"></script>
  </body>
</html>